### PR TITLE
Direct hyperlinks

### LINKED ISSUE #10: Find a representation for relative URLs.
While working on hyperlink references as part of issue #7, I realized that we need a way to represent relative URLs in the RST doctree. The [servo/rust-url](https://github.com/servo/rust-url) crate we're using can parse relative URL strings with a base, but can only store absolute URLs. Unfortunately, looking through the first pages of crates.io, nobody seems to have tried to reinvent the wheel in that matter.

For the time being, I'm leaving relative URL references unimplemented. In the meantime I'll fork and play around with the url crate to try and make a version that can represent relative URLs.

### PR DESCRIPTION
**Pull request by andreubotella**
This is a complete implementation of direct hyperlinks, including standalone URLs, with parse rules for indirect hyperlinks. This closes #10 and helps set the stage for some of the items in #7.

As in #9, I didn't bother to update the HTML output in this PR. After we're done or mostly done with the references, I'll be working on getting the renderer up to date.

### CODE REVIEW

File: src/rst.pest [hunk 1]
```diff
@@ -285,7 +355,7 @@ marker = _{ (bullet_marker | "..") ~ " " }
 // star_line = { "****" ~ "*"* | spacechar ~ "*"+ ~ &spacechar }
 // ul_line = { "____" ~ "_"* | spacechar ~ "_"+ ~ &spacechar }
 
-// whitespace = { spacechar | NEWLINE }
+whitespace = { spacechar | NEWLINE }
```
Comment (flying-sheep) [inline src/rst.pest]: Neither whitespace nor spacechar later are used in other rules, and I hope they don’t need to be. Please leave them commented.

```suggestion
// whitespace = { spacechar | NEWLINE }
```

File: src/rst.pest [hunk 2]
```diff
@@ -128,8 +128,11 @@ str = { (!(NEWLINE | reference | substitution_ref) ~ ANY)+ }
 reference = { reference_target | reference_explicit | reference_auto }
 
 reference_target = { reference_target_uq ~ "_" | reference_target_qu }
-reference_target_uq =  { (!("_"|":"|"`") ~ nonspacechar)* }
-reference_target_qu = _{ ( !("`"? ~ "`_") ~ "`" ~ !"``" ) ~ target_name_qu ~ ( "`" ~ !"``" ) ~ "_" }
+reference_target_uq =  { (!("_"|":"|"`") ~ nonspacechar)+ }
+reference_target_qu = { ( !("`"? ~ "`_") ~ "`" ~ !"``" ) ~ reference_text? ~ ("<" ~ reference_bracketed ~ ">")? ~ ( "`" ~ !"``" ) ~ "_" }
+reference_text = { !("<") ~ ( !("`"|"<") ~ ANY )+ }
```
Comment (flying-sheep) [inline src/rst.pest]:
```suggestion
reference_text = { !"<" ~ ( !("`"|"<") ~ ANY )+ }
```

File: src/rst.pest [hunk 3]
```diff
@@ -308,7 +378,7 @@ marker = _{ (bullet_marker | "..") ~ " " }
 //     "\"" ~ (!"\"" ~ ANY)* ~ "\"" |
 //     "'"  ~ (!"'"  ~ ANY)* ~ "'"
 // }
-// spacechar = _{ " " | "\t" }
+spacechar = _{ " " | "\t" }
```
Comment (flying-sheep) [inline src/rst.pest]:
```suggestion
// spacechar = _{ " " | "\t" }
```

File: src/rst.pest [hunk 4]
```diff
@@ -139,14 +142,81 @@ reference_title        = { ( reference_title_single | reference_title_double | "
...
 
-reference_auto = { reference_embedded | reference_auto_url | reference_auto_email }
-reference_embedded = { "`" ~ reference_embedded_source ~ "<" ~ ASCII_ALPHA+ ~ "://" ~ (!(NEWLINE|">") ~ ANY)+ ~ ">`_" ~ "_"? }
-reference_embedded_source = { ( !("<"|":"|"`") ~ ( " " | nonspacechar | blank_line ) )* }
-reference_auto_url = { ASCII_ALPHA+ ~ "://" ~ (!(NEWLINE|">") ~ ANY)+ }
-reference_auto_email = { "<" ~ "mailto:"? ~ (ASCII_ALPHANUMERIC|"-"|"+"|"_"|"."|"/"|"!"|"%"|"~"|"$")+ ~ "@" ~ (!(NEWLINE | ">") ~ ANY)+ ~ ">" }
+// Emails can't end with punctuation, but URLs must use a separate rule.
+reference_auto = { url_auto | email }
+//reference_embedded = { "`" ~ reference_embedded_source ~ "<" ~ absolute_url_with_fragment ~ ">`_" ~ "_"? }
+//reference_embedded_source = { ( !("<"|":"|"`") ~ ( " " | nonspacechar | blank_line ) )* }
```
Comment (flying-sheep) [inline src/rst.pest]: Why did you remove those? Shouldn’t we just panic with an “unimplemented” message if we encounter these?

File: src/rst.pest (hunk 1 above)
Comment (andreubotella) [reply src/rst.pest]: My bad, I tried using the whitespace rule to deal with the space between the reference text and the bracketed target, before realizing that `` `<target_only>`_`` is a valid reference, so I just trimmed the space in the conversion.

File: src/rst.pest (hunk 4 above)
Comment (andreubotella) [reply src/rst.pest]: I removed it because I'd changed `reference_target_qu` to handle all references in backticks, including those with a bracketed target. Since I'd defined the rules for URL parsing separately, it made sense to merge both those rules and add support for references like `` `<target_only>`_``.

**Comment by flying-sheep** [comment]
Merged in 5387291c1a2d4cfd0e5acdad26dcc7e33329d39a..df6b7645d845a022ca7eeba08b1ecb761a020195
